src/WinRT.Runtime/Interop/IContextCallback.cs

```diff
@@ -18,11 +19,36 @@ internal struct ComCallData
     }
 
 #if NET && CsWinRT_LANG_11_FEATURES
-    internal unsafe struct CallbackData
+    internal sealed unsafe class CallbackData
     {
+        [ThreadStatic]
+        private static CallbackData TlsInstance;
+
         public delegate*<object, void> Callback;
         public object State;
+        public GCHandle Handle;
+
+        private CallbackData()
+        {
+            // Create a handle to access the object from a native callback invoked on another thread.
+            // The handle is weak to ensure that the object does not leak (or it would keep itself
+            // alive). The target is guaranteed to be alive because callers will use 'GC.KeepAlive'.
+            Handle = GCHandle.Alloc(this, GCHandleType.Weak);
+        }
+
+        ~CallbackData()
+        {
+            Handle.Free();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static CallbackData GetOrCreate()

```

Comment:
Are you sure that there cannot ever be recursion that would cause multiple recursive uses of CallbackData on the same thread?

Calls between STA apartments always come with a lot of surprises and corner-case bugs...

Reply:
Applied the suggested changes and also left some notes about reentrancy. I talked with @manodasanW as well and in this scenario this should be fine. We're only ever doing this context switch to release an object reference on its original context, and nothing else. There is no way that should/could then somehow switch back to this starting thread (it wouldn't even know which one that was) to then try to schedule another context callback recursively. So this should be fine. Also in ASTA scenarios this would also just be flat out impossible by construction (because ASTA blocks reentrant calls into it, so even if hypothetically we somehow ended up there, that second context callback would just throw before being scheduled). But then again either case should just never happen. Definitely a good point though and worth writing some additional comments about this! 😄